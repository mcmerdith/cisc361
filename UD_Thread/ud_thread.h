/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
