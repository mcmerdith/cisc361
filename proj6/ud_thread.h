/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;