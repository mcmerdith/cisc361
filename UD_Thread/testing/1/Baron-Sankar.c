/* Lauren Baron and Swetha Sankar
 * CISC361-010
 * Programming Assignment 6: User-Level Thread Library
 * Phase 1
 * Extra Credit: Write our own challenging test case
 * NOTE: takes user input in order to know how many threads to create
 * to get the output we got in our ECOutput.txt file, input 6
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "ud_thread.h"

int threadNum;
#define BUFFER_SIZE 256

void my_thread(int id)
{
	for (int i = 0; i < threadNum; i++)
	{
		printf("Thread %d yielding\n", id);
		t_yield();
	}
	t_terminate();
	printf("Thread %d terminated\n", id);
}

int main(void)
{
	printf("\n\n");
	// ask user to input how many threads they want to create
	char tempNum[BUFFER_SIZE];
	printf("Enter number of threads: ");
	fgets(tempNum, BUFFER_SIZE, stdin);

	printf("\n\n");
	// init
	threadNum = atoi(tempNum);
	printf("Initializing user-level thread library\n");
	t_init();
	printf("User-level thread library initialized\n");

	printf("\n\n"); // spacer

	// create
	printf("Creating %d threads\n", threadNum);
	for (int i = 0; i < threadNum + 1; i++)
	{
		t_create(my_thread, i, 1);
	}
	printf("Created %d threads\n", threadNum);

	printf("\n\n"); // spacer

	// yield
	for (int i = 0; i < threadNum; i++)
	{
		printf("Thread %d yielding\n", i + 1);
		t_yield();
		printf("\n");
	}

	// terminate by calling yield one more time
	printf("\n\n");
	printf("Terminating threads...\n");
	t_yield();

	// shut down
	printf("\n\n");
	printf("Shutting down user-level thread library\n");
	t_shutdown();
	printf("User-level thread library shut down\n");

	printf("\n\n");
	return 0;
}
