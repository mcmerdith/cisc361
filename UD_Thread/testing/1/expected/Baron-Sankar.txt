

Enter number of threads: 6


Initializing user-level thread library
User-level thread library initialized


Creating 6 threads
Created 6 threads


Thread 1 yielding
Thread 0 yielding
Thread 1 yielding
Thread 2 yielding
Thread 3 yielding
Thread 4 yielding
Thread 5 yielding
Thread 6 yielding

Thread 2 yielding
Thread 0 yielding
Thread 1 yielding
Thread 2 yielding
Thread 3 yielding
Thread 4 yielding
Thread 5 yielding
Thread 6 yielding

Thread 3 yielding
Thread 0 yielding
Thread 1 yielding
Thread 2 yielding
Thread 3 yielding
Thread 4 yielding
Thread 5 yielding
Thread 6 yielding

Thread 4 yielding
Thread 0 yielding
Thread 1 yielding
Thread 2 yielding
Thread 3 yielding
Thread 4 yielding
Thread 5 yielding
Thread 6 yielding

Thread 5 yielding
Thread 0 yielding
Thread 1 yielding
Thread 2 yielding
Thread 3 yielding
Thread 4 yielding
Thread 5 yielding
Thread 6 yielding

Thread 6 yielding
Thread 0 yielding
Thread 1 yielding
Thread 2 yielding
Thread 3 yielding
Thread 4 yielding
Thread 5 yielding
Thread 6 yielding



Terminating threads...


Shutting down user-level thread library
User-level thread library shut down


